### PR TITLE
BartForCausalLM analogs to `ProphetNetForCausalLM`

### LINKED ISSUE #9066: Add BartForCausalLM analogs to `ProphetNetForCausalLM`
# 🚀 Feature request

Bart is a seq2seq model, but there might be applications where one would like to use only the pre-trained BartDecoder in an EncoderDecoder setting with a "long" encoder, such as 

```python

from transformers import EncoderDecoderModel

model = EncoderDecoderModel("allenai/longformer-large-4096", "facebook/bart-large")

# fine-tune model ...
```

This is already possible for ProphetNet:

```python
from transformers import EncoderDecoderModel
import torch

model = EncoderDecoderModel.from_encoder_decoder_pretrained("allenai/longformer-large-4096", "microsoft/prophetnet-large-uncased")

input_ids = torch.tensor([10 * [1]])
labels = torch.tensor([10 * [0]])
loss = model(input_ids, decoder_input_ids=labels, labels=labels).loss
loss.backward()
```

, but not yet for Bart. This "Good first/second issue" is about implemented a `BartForCausalLM` analogs to the one in ProphetNet here:
https://github.com/huggingface/transformers/blob/9cc9f4122e2a1027a6011951e3c6629a0f1b6c3e/src/transformers/models/prophetnet/modeling_prophetnet.py#L1882

To verify that the feature works as expected, one should make sure that the following tests are added:
- A `BartStandaloneDecoderModelTest` class as is done in https://github.com/huggingface/transformers/blob/9cc9f4122e2a1027a6011951e3c6629a0f1b6c3e/tests/test_modeling_prophetnet.py#L1072
- And an encoder-decoder test class as it's done here: https://github.com/huggingface/transformers/blob/9cc9f4122e2a1027a6011951e3c6629a0f1b6c3e/tests/test_modeling_encoder_decoder.py#L758

## Motivation

mentioned above for Long-range seq2seq warm-starting e.g.

## Your contribution

I'm more than happy to guide someone through this issue!
It's a bit more advanced so I'll give it both "Good first issue" and "Good second issue".

You can claim the issue by writing it below and/or opening a PR :-)

### PR DESCRIPTION
**Pull request by sadakmed**
# What does this PR do?

Implementing BartForCausalLM anologs for ProphetNetForCausalLM

Fixes #9066

### CODE REVIEW

**Comment by patrickvonplaten** [comment]
Hy @sadakmed, 

let me know if you need help on the issue or if you don't find the time to tackle it. I'll then just make it open to the "public" again :-)

**Comment by sadakmed** [comment]
> Hy @sadakmed,
> 
> let me know if you need help on the issue or if you don't find the time to tackle it. I'll then just make it open to the "public" again :-)

I'm in a vacation now, if it's so urgent you can make it public, else I will get back to it as soon as possible

File: src/transformers/models/bart/modeling_bart.py [hunk 1]
```diff
@@ -1518,3 +1520,230 @@ def forward(
...
+            return_dict=return_dict,
+        )
+
+        batch_size, sequence_length = input_ids.shape if input_ids is not None else inputs_embeds.shape[:2]
+
+        predicting_streams = outputs[1].view(batch_size, self.config.ngram, sequence_length, -1)
+        predict_logits = self.lm_head(predicting_streams)
+
+        logits = predict_logits[:, 0]
+        logits_ngram = predict_logits[:, 1:] if self.config.ngram > 1 else None
```
Comment (patrickvonplaten) [inline src/transformers/models/bart/modeling_bart.py:1608]: Bart has no `config.ngram` parameter. Note that the loss can be computed as easily as it is done in `BartForConditionalGeneration`: https://github.com/huggingface/transformers/blob/a24e1250cedd8eb897d3e42b9c8eef8dde15c184/src/transformers/models/bart/modeling_bart.py#L1254

File: src/transformers/models/bart/modeling_bart.py [hunk 2]
```diff
@@ -1518,3 +1520,230 @@ def forward(
...
+        # first step, decoder_cached_states are empty
+        return {
+            "input_ids": input_ids,  # encoder_outputs is defined. input_ids not needed
+            "attention_mask": attention_mask,
+            "past_key_values": past,
+            "use_cache": use_cache,
+        }
+
+    @staticmethod
+    def _reorder_cache(past, beam_idx):
```
Comment (patrickvonplaten) [inline src/transformers/models/bart/modeling_bart.py:1672]: this should be a copy of this function: https://github.com/huggingface/transformers/blob/a24e1250cedd8eb897d3e42b9c8eef8dde15c184/src/transformers/models/bart/modeling_bart.py#L1300

File: src/transformers/models/bart/modeling_bart.py [hunk 3]
```diff
@@ -1518,3 +1520,230 @@ def forward(
...
+            batch_size, num_attn_heads, decoder_sequence_length, embed_size_per_head)`).
+
+            Contains pre-computed hidden-states (key and values in the attention blocks) of the decoder that can be
+            used (see :obj:`past_key_values` input) to speed up sequential decoding.
+        hidden_states (:obj:`tuple(torch.FloatTensor)`, `optional`, returned when ``output_hidden_states=True`` is passed or when ``config.output_hidden_states=True``):
+            Tuple of :obj:`torch.FloatTensor` (one for the output of the embeddings + one for the output of each layer)
+            of shape :obj:`(batch_size, decoder_sequence_length, hidden_size)`.
+
+            Hidden-states of main stream of the decoder at the output of each layer plus the initial embedding outputs.
+        ngram_hidden_states (:obj:`tuple(torch.FloatTensor)`, `optional`, returned when ``output_hidden_states=True`` is passed or when ``config.output_hidden_states=True``):
```
Comment (patrickvonplaten) [inline src/transformers/models/bart/modeling_bart.py:1715]: bart has no `ngram_....` outputs => The dataclass looks good, we just need to delete `logits_ngram` and `ngram_hidden_states` and `ngram_attentions`

**Review by patrickvonplaten — COMMENTED** [review]
It's great to see progress here! Good job. Note that `BartForCausalLM` should have the exact same functionality in terms of computing the loss as the `BartForConditionalGeneration` model => this means we can copy the way `BartForConditionalGeneration` computes the loss into `BartForCausalLM`. 

Let me know if you need help or are stuck :-)

**Comment by sadakmed** [comment]
@patrickvonplaten  The loss function it what I stuck on, thank you very much for your guidance.


> Let me know if you need help or are stuck :-)

for sure I will ;-)

**Comment by patrickvonplaten** [comment]
Hey @sadakmed,

Do you have an update on the PR? It's been three weeks now and it would be great to merge this soon. Sorry, we're very fast-moving in this lib and other community contributors have started asking for this feature. By next week, I'll probably have to take a look myself or redistribute the issue.

File: src/transformers/models/bart/modeling_bart.py (hunk 2 above)
Comment (sadakmed) [reply src/transformers/models/bart/modeling_bart.py:1672]: done

**Comment by sadakmed** [comment]
Hi @patrickvonplaten  my apologies, 

Could you Please see it now, lemme know if anything is missing.

**Comment by sadakmed** [comment]
Hi @patrickvonplaten, working on the test 'BartStandaloneCausalLM': I dont know if the `self.model_tester` in 'setUp' should be 'BartDecoderTester' (needed to be implemented), or do u recommend something else.

**Comment by patrickvonplaten** [comment]
Hey @sadakmed, 

Thanks a lot for you additions here :-) 
Yes, we need a new `BartStandaloneDecoderModelTester` analog to how it's done for ProphetNet in `tests/test_modeling_prophetnet.py`. Do you want to give it a try? Otherwise, I can go into your PR and see how to add the tests :-)

**Comment by sadakmed** [comment]
> Do you want to give it a try?

of course, I'm working on it,

**Comment by sadakmed** [comment]
Hi @patrickvonplaten, I just pushed the test, could you please check it out! 
thaaaanks

**Comment by patrickvonplaten** [comment]
Hey @sadakmed, 

I corrected `BartForCausalLM` and also added `MBartForCausalLM`. It would be awesome if you could take care of adding `MarianForCausalLM`, `PegasusForCausalLM`, `BlenderbotForCausalLM`, and `BlenderbotSmallForCausalLM`. 

To do so you can simply copy everything that was done for `MBart` in this PR 1-to-1 to the mentioned models above. Let me know if sounds feasible for you :-) 

Thanks a lot for your help so far!

**Comment by sadakmed** [comment]
> adding `MarianForCausalLM`, `PegasusForCausalLM`, `BlenderbotForCausalLM`, and `BlenderbotSmallForCausalLM`.

implementing it for use or for test? 
Yes I would like to do it, with all pleasure.

**Comment by patrickvonplaten** [comment]
> > adding `MarianForCausalLM`, `PegasusForCausalLM`, `BlenderbotForCausalLM`, and `BlenderbotSmallForCausalLM`.
> 
> implementing it for use with EncoderDecoder or just the test?
> 
> Yes I would like to do it, with all pleasure.

For those models, there is no need to add a test to `EncoderDecoderModel`. We should only copy-paste the code that was added to MBart to those models and also copy-paste the test in `test_modeling_marian.py` e.g.

**Comment by sadakmed** [comment]
@patrickvonplaten  Could you please check the test if it well, and about the test of `Decoder only` I didn't get what do you mean!!

File: tests/test_modeling_marian.py [hunk 4]
```diff
@@ -538,3 +538,221 @@ def test_undoing_renaming(self):
...
+
+        inputs_dict = {
+            "input_ids": input_ids,
+            "attention_mask": attention_mask,
+        }
+        return config, inputs_dict
+
+
+@require_torch
+class MBartStandaloneDecoderModelTest(ModelTesterMixin, GenerationTesterMixin, unittest.TestCase):
```
Comment (patrickvonplaten) [inline tests/test_modeling_marian.py:730]: should be changed to `MarianStandaloneDecoderModelTest`

File: tests/test_modeling_marian.py [hunk 5]
```diff
@@ -538,3 +538,221 @@ def test_undoing_renaming(self):
             "en-de",
         ]
         self.assertListEqual(expected_opus_names, converted_opus_names)
+
+class MBartStandaloneDecoderModelTester:
```
Comment (patrickvonplaten) [inline tests/test_modeling_marian.py:542]:
```suggestion
class MarianStandaloneDecoderModelTester:
```

**Comment by patrickvonplaten** [comment]
It would be nice to fix the tests and also add tests for `Pegasus`, `Blenderbot`, and `BlenderbotSmall`

**Comment by sadakmed** [comment]
> It would be nice to fix the tests and also add tests for `Pegasus`, `Blenderbot`, and `BlenderbotSmall`

exactly like the one was for Marian?

**Comment by sadakmed** [comment]
@patrickvonplaten  could you check please!

File: src/transformers/models/bart/modeling_bart.py [hunk 6]
```diff
@@ -843,6 +843,30 @@ def __init__(self, config: BartConfig, embed_tokens: Optional[nn.Embedding] = No
 
         self.init_weights()
 
+    def get_input_embeddings(self):
+        return self.embed_tokens
+
+    def set_input_embeddings(self, value):
+        self.embed_tokens = value
+
+    def _prepare_decoder_attention_mask(self, attention_mask, input_shape, inputs_embeds, past_key_values_length):
```
Comment (patrickvonplaten) [inline src/transformers/models/bart/modeling_bart.py:852]: This both corrects a bug and refactors this rather complex function so that it can be copied into all Bart-ilke models

File: utils/check_repo.py [hunk 7]
```diff
@@ -205,9 +210,8 @@ def find_tested_models(test_file):
     with open(os.path.join(PATH_TO_TESTS, test_file), "r", encoding="utf-8", newline="\n") as f:
         content = f.read()
     all_models = re.findall(r"all_model_classes\s+=\s+\(\s*\(([^\)]*)\)", content)
-    # Check with one less parenthesis
-    if len(all_models) == 0:
-        all_models = re.findall(r"all_model_classes\s+=\s+\(([^\)]*)\)", content)
+    # Check with one less parenthesis as well
+    all_models += re.findall(r"all_model_classes\s+=\s+\(([^\)]*)\)", content)
```
Comment (patrickvonplaten) [inline utils/check_repo.py:214]: This failed since some model testers have both cases - hope this change is ok @sgugger

**Comment by patrickvonplaten** [comment]
**UPDATE:**

@LysandreJik @sgugger 
This PR enables all Bart-like models to be used in combination with the Encoder-Decoder framework. The model `BartForCausalLM` is added for Bart and then copied to all other models via the copying mechanism. Also, a new model tester is added for all those models. 
While working on this I found a small bug for a very edge-case scenario for Bart and corrected it here: https://github.com/huggingface/transformers/pull/9128/files#r569436360 . The newly added tests were failing, which made me aware of the bug. 
Also, I had to slightly change the `check_repo.py` file so that it counts both classes from `all_model_classes` with 1 and 2 paratheses.

File: src/transformers/models/blenderbot/modeling_blenderbot.py [hunk 8]
```diff
@@ -1336,3 +1352,193 @@ def _reorder_cache(past, beam_idx):
...
+    def __init__(self, config):
+        super().__init__(config)
+        self.decoder = BlenderbotDecoder(config)
+
+    def forward(self, *args, **kwargs):
+        return self.decoder(*args, **kwargs)
+
+
+# Copied from transformers.models.bart.modeling_bart.BartForCausalLM with Bart->Blenderbot
+class BlenderbotForCausalLM(BlenderbotPreTrainedModel):
```
Comment (sgugger) [inline src/transformers/models/blenderbot/modeling_blenderbot.py:1373]: :heart: :heart: :heart: :heart: :heart:

File: src/transformers/models/prophetnet/modeling_prophetnet.py [hunk 9]
```diff
@@ -1330,7 +1330,7 @@ def forward(
             >>> import torch
 
             >>> tokenizer = ProphetNetTokenizer.from_pretrained('microsoft/prophetnet-large-uncased')
-            >>> model = ProphetNetDecoder.from_pretrained('patrickvonplaten/prophetnet-large-uncased-standalone', add_cross_attention=False)
+            >>> model = ProphetNetDecoder.from_pretrained('microsoft/prophetnet-large-uncased', add_cross_attention=False)
```
Comment (sgugger) [inline src/transformers/models/prophetnet/modeling_prophetnet.py:1333]: Nice catch!

File: utils/check_repo.py (hunk 7 above)
Comment (sgugger) [reply utils/check_repo.py:214]: Should be fine, it seems like the right fix to me!

File: utils/check_repo.py [hunk 10]
```diff
@@ -32,17 +32,17 @@
...
+    "BlenderbotSmallDecoderWrapper",  # Building part of bigger (tested) model.
     "BlenderbotEncoder",  # Building part of bigger (tested) model.
-    "BlenderbotDecoder",  # Building part of bigger (tested) model.
+    "BlenderbotDecoderWrapper",  # Building part of bigger (tested) model.
     "MBartEncoder",  # Building part of bigger (tested) model.
-    "MBartDecoder",  # Building part of bigger (tested) model.
+    "MBartDecoderWrapper",  # Building part of bigger (tested) model.
     "PegasusEncoder",  # Building part of bigger (tested) model.
-    "PegasusDecoder",  # Building part of bigger (tested) model.
+    "PegasusDecoderWrapper",  # Building part of bigger (tested) model.
```
Comment (sgugger) [inline utils/check_repo.py:45]: Should really ignore all models ending with `Encoder`/`Decoder` and now `DecoderWrapper` ;-)

**Review by sgugger — APPROVED** [review]
This is fantastic work! Thanks a lot!

File: tests/test_modeling_pegasus.py [hunk 11]
```diff
@@ -308,3 +307,224 @@ def test_pegasus_xsum_summary(self):
...
+        return config, inputs_dict
+
+
+@require_torch
+class PegasusStandaloneDecoderModelTest(ModelTesterMixin, GenerationTesterMixin, unittest.TestCase):
+    all_model_classes = (PegasusDecoder, PegasusForCausalLM) if is_torch_available() else ()
+    all_generative_model_classes = (PegasusForCausalLM,) if is_torch_available() else ()
+    test_pruning = False
+    test_torchscript = False
+    test_resize_embeddings = False
```
Comment (LysandreJik) [inline tests/test_modeling_pegasus.py:507]: Will resizing the token embeddings be allowed at one point? Seems like something that could be done easily given other BART-like models have this functionality.

File: templates/adding_a_new_model/cookiecutter-template-{{cookiecutter.modelname}}/modeling_{{cookiecutter.lowercase_modelname}}.py [hunk 12]
```diff
@@ -1924,7 +1925,7 @@ def forward(
         return outputs
 
 
-# Copied from transformers.models.bart.modeling_bart.BartClassificationHead with Bart->{{cookiecutter.camelcase_modelname}}
+# Copied from transformers.models.model.modeling_model.BartClassificationHead with Bart->{{cookiecutter.camelcase_modelname}}
```
Comment (LysandreJik) [inline templates/adding_a_new_model/cookiecutter-template-{{cookiecutter.modelname}}/modeling_{{cookiecutter.lowercase_modelname}}.py:1928]: Is this change intentional?

File: templates/adding_a_new_model/cookiecutter-template-{{cookiecutter.modelname}}/modeling_{{cookiecutter.lowercase_modelname}}.py [hunk 13]
```diff
@@ -2073,6 +2074,35 @@ def dummy_inputs(self):
 """
 
 
+{{cookiecutter.uppercase_modelname}}_STANDALONE_INPUTS_DOCSTRING = r"""
+    Args:
+        input_ids (:obj:`torch.LongTensor` of shape :obj:`(batch_size, sequence_length)`):
+            Indices of input sequence tokens in the vocabulary. Padding will be ignored by default should you provide
+            it.
+
+            Indices can be obtained using :class:`~transformers.ProphetNetTokenizer`. See
```
Comment (LysandreJik) [inline templates/adding_a_new_model/cookiecutter-template-{{cookiecutter.modelname}}/modeling_{{cookiecutter.lowercase_modelname}}.py:2083]: Should not mention ProphetNet here

File: templates/adding_a_new_model/cookiecutter-template-{{cookiecutter.modelname}}/modeling_{{cookiecutter.lowercase_modelname}}.py [hunk 14]
```diff
@@ -2919,4 +2955,163 @@ def forward(
             encoder_hidden_states=outputs.encoder_hidden_states,
             encoder_attentions=outputs.encoder_attentions,
         )
+
+# Copied from transformers.models.model.modeling_model.BartDecoderWrapper with Bart->{{cookiecutter.camelcase_modelname}}
```
Comment (LysandreJik) [inline templates/adding_a_new_model/cookiecutter-template-{{cookiecutter.modelname}}/modeling_{{cookiecutter.lowercase_modelname}}.py:2959]: Same here, should it be `model.modeling_model`?

File: templates/adding_a_new_model/cookiecutter-template-{{cookiecutter.modelname}}/modeling_{{cookiecutter.lowercase_modelname}}.py [hunk 15]
```diff
@@ -2919,4 +2955,163 @@ def forward(
...
+
+    def __init__(self, config):
+        super().__init__(config)
+        self.decoder = {{cookiecutter.camelcase_modelname}}Decoder(config)
+
+    def forward(self, *args, **kwargs):
+        return self.decoder(*args, **kwargs)
+
+
+# Copied from transformers.models.model.modeling_model.BartForCausalLM with Bart->{{cookiecutter.camelcase_modelname}}
```
Comment (LysandreJik) [inline templates/adding_a_new_model/cookiecutter-template-{{cookiecutter.modelname}}/modeling_{{cookiecutter.lowercase_modelname}}.py:2974]: Here too

**Review by LysandreJik — APPROVED** [review]
Impressive work, great work @sadakmed and @patrickvonplaten!

Very cool that you completed the templates as well.

There seems to be a few issues with the "copied by" scripts in the templates, but the test still passes, which is weird.

File: tests/test_modeling_pegasus.py (hunk 11 above)
Comment (patrickvonplaten) [reply tests/test_modeling_pegasus.py]: good catch - enabled all those tests

**Comment by patrickvonplaten** [comment]
Great job @sadakmed

**Comment by sadakmed** [comment]
> Great job @sadakmed

wouldn't happen without you, thank you very much.  see u in the next PR ;)
